Source/Core/Core/State.cpp

```diff
@@ -121,20 +120,31 @@ static const std::map<u32, std::pair<std::string, std::string>> s_old_versions =
     {38, {"4.0-4963", "4.0-5267"}}, {39, {"4.0-5279", "4.0-5525"}}, {40, {"4.0-5531", "4.0-5809"}},
     {41, {"4.0-5811", "4.0-5923"}}, {42, {"4.0-5925", "4.0-5946"}}};
 
-enum
-{
-  STATE_NONE = 0,
-  STATE_SAVE = 1,
-  STATE_LOAD = 2,
-};
+static constexpr bool s_use_compression = true;
 
-static bool s_use_compression = true;
+// Acquired when starting a task that will write state saves to the filesystem.
+// It's not a real mutex, it can be moved to other threads.
+[[nodiscard]] static auto GetStateSaveTaskLock()
+{
+  return std::unique_lock{s_state_saves_in_progress};
+}
 
-void EnableCompression(bool compression)
+// Waits for previously requested state saves to reach the filesystem.
+static void WaitForStateSaveTaskCompletion()
 {
-  s_use_compression = compression;
+  // Note: Because CPU-thread might be in a task itself, we instead wait for the worker.
+  // Being the CPU-thread, we would have already queued previous work.
+  if (Core::IsCPUThread())
+    s_state_write_thread.WaitForCompletion();
+  else
+    s_state_saves_in_progress.WaitForValue(0);
 }
 
+static Common::EventHook s_flush_data_hook =
+    Core::FlushUnsavedDataEvent::Register(&WaitForStateSaveTaskCompletion, "State flush");
```

Comment:
"Savestate flush" or "State save flush" would be clearer. More or less everything that registers for this event would be flushing some type of state (e.g. GC memory card state).